File: ts/query.ts
import * as Parser from "./parser/parser";
import * as Planner from "./planner/planner";
import * as Executor from "./executor/executor";
import { Iter } from "./executor/producer";

/**
 * Coordinates the evaluation of a query to an iterator. 
 * 
 * @param input A SELECT query or a sp_ stored procedure.
 * @returns An iterator
 */
export async function evaluate(input: string): Promise<Iter> {
    const command: Parser.Command = Parser.parse(input);
    const plan: Planner.Plan = await Planner.plan(command);
    return Executor.run(plan);
}

File: rust/src/query.rs
use serde_json::Value;

mod producer;
mod filter;
mod parser;
mod planner;
mod executor;

pub fn evaluate(user_input: &str) -> Box<dyn Itrator> {
    let query = parser::parse(user_input.to_string());
    let plan: planner::Plan = planner::plan(query);
    executor::run(plan)
}

pub type Row = Vec<Value>;
pub trait Itrator {
    fn next(&mut self) -> Option<Row>;
}